ink-wrapper-types/src/drink/client.rs

```diff
@@ -0,0 +1,165 @@
+    {
+        let actor = self.get_actor();
+        let gas_limit = self.get_gas_limit();
+
+        let instantiate_contract_result = self.sandbox().instantiate_contract(
+            call.code_hash.to_vec(),
+            call.value,
+            call.data,
+            call.salt,
+            actor,
+            gas_limit,
+            None,
+        );
+
+        let contract_address = match &instantiate_contract_result.result {
+            Ok(exec_result) if exec_result.result.did_revert() => Err(Error::DeploymentReverted),
+            Err(err) => Err(Error::DeploymentFailed(*err)),
+            Ok(exec_result) => Ok(exec_result.account_id.clone()),
+        }?;
+
+        let events = extract_events(&instantiate_contract_result.events);
+
+        Ok(ContractInstantiateResult {
+            gas_consumed: instantiate_contract_result.gas_consumed,
+            gas_required: instantiate_contract_result.gas_required,
+            result: contract_address,
+            events,
+        })
+    }
+
+    fn execute<T: scale::Decode + Send + std::fmt::Debug>(
+        &mut self,
+        call: ExecCall<T>,
+    ) -> Result<ContractExecResult<T>, Error> {
+        let actor = self.get_actor();
+        let gas_limit = self.get_gas_limit();
+        let contract_address = (*AsRef::<[u8; 32]>::as_ref(&call.account_id)).into();
+
+        let result = call_contract(
+            actor,
+            gas_limit,
+            self.sandbox(),
+            contract_address,
+            call.value,
+            call.data,
+        )?;
+
+        Ok(result)
+    }
+
+    fn query<T: scale::Decode + Send + std::fmt::Debug>(
+        &mut self,
+        call: ReadCall<T>,
+    ) -> Result<ContractReadResult<T>, Error> {
+        let actor = self.get_actor();
+        let gas_limit = self.get_gas_limit();
+        let contract_address = (*AsRef::<[u8; 32]>::as_ref(&call.account_id)).into();
+
+        let result = self.sandbox().dry_run(|sandbox| {
+            call_contract(
+                actor,
+                gas_limit,
+                sandbox,
+                contract_address,
+                call.value,
+                call.data,
+            )
+        })?;
+
+        Ok(result)
+    }
+}
+
+fn call_contract<T: scale::Decode + Send + std::fmt::Debug>(
+    actor: <MinimalRuntime as frame_system::Config>::AccountId,
+    gas_limit: Weight,
+    sandbox: &mut drink::Sandbox<MinimalRuntime>,
+    address: <MinimalRuntime as frame_system::Config>::AccountId,
+    value: u128,
+    data: Vec<u8>,
+) -> Result<ContractResult<T>, Error> {
+    let result = sandbox.call_contract(
+        address,
+        value,
+        data,
+        actor,
+        gas_limit,
+        None,
+        pallet_contracts::Determinism::Enforced,
+    );
+
+    let message_result: T = match &result.result {
+        // NOTE to reviewers: we're not checking if the contract call reverted here.
+        // We will most probably return an error from the contract call anyway.
```

Comment:
```suggestion
```

Reply:
Not sure if this will bite us.

Reply:
I've added the `revered` bool to the result https://github.com/Cardinal-Cryptography/ink-wrapper/pull/68/commits/a2df74c82e15089262c3fba8d288dead522dc157. Please take a look.